src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java

```diff
@@ -407,10 +440,61 @@ private void emitSetOutput(Class<?> storeType) {
     private void emitGetInput() {
         Class<?> highLevelType = callerMethodType.parameterType(paramIndex);
         emitLoad(BasicType.of(highLevelType), paramIndex2ParamSlot[paramIndex]);
+
+        if (shouldAcquire(highLevelType)) {
+            emitDup(BasicType.L);
+            emitAcquireScope();
+        }
+
         pushType(highLevelType);
         paramIndex++;
     }
 
+    private void emitAcquireScope() {
+        emitCheckCast(Scoped.class);
+        emitInvokeInterface(Scoped.class, "scope", SCOPE_DESC);
+        Label skipAcquire = new Label();
+        Label end = new Label();
+
+        // start with 1 scope to maybe acquire on the stack
+        assert curScopeLocalIdx != -1;
+        boolean hasOtherScopes = curScopeLocalIdx != 0;
+        for (int i = 0; i < curScopeLocalIdx; i++) {
+            emitDup(BasicType.L); // dup for comparison
+            emitLoad(BasicType.L, scopeSlots[i]);
+            mv.visitJumpInsn(IF_ACMPEQ, skipAcquire);
+        }
+
+        // 1 scope to acquire on the stack
+        emitDup(BasicType.L);
+        int nextScopeLocal = scopeSlots[curScopeLocalIdx++];
+        emitStore(BasicType.L, nextScopeLocal); // store off one to release later
+        emitCheckCast(ResourceScopeImpl.class);
+        emitInvokeVirtual(ResourceScopeImpl.class, "acquire0", ACQUIRE0_DESC); // call acquire on the other
+
+        if (hasOtherScopes) { // avoid ASM generating a bunch of nops for the dead code
+            mv.visitJumpInsn(GOTO, end);
+
+            mv.visitLabel(skipAcquire);
+            mv.visitInsn(POP); // drop scope
+        }
+
+        mv.visitLabel(end);
+    }
+
+    private void emitReleaseScopes() {
+        for (int scopeLocal : scopeSlots) {
```

Comment:
Isn't this emitting more code than required? E.g. if we pass 10 primitives, we will emit 10 scope checks at the end of the adapter? Granted, since all these locals contain "null", there's no correctness issue here - but it is odd to (i) have a local for a scope corresponding to a primitive type (even though we know we never need it) and have extra logic in the preamble/postamble because of it.

Reply:
There is only a scope slot local for Addressable and NativeSymbol parameters. See the logic that initializes `scopeSlots`:

```java
scopeSlots = Arrays.copyOf(scopeSlots, numScopes); // fit to size
```

(the initial array is the size of the parameter list initially, but is adjusted downwards. Maybe that should be made clearer...)

Reply:
I've tried to clarify the code that creates `scopeSlots` a bit by using a local variable for the intermediate array, instead of storing it in the field.

Reply:
Thanks - I've missed that